File: apps/overlay-windows/src/PrismAuth.cs
using System.Net.Http;
using System.Text.Json;

namespace PrismOverlay;

/// CLI-style device pairing: opens the Prism /link page in the browser, the user
/// signs in or creates an account, and the device key is fetched and saved.
public sealed class PrismAuth
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(8) };
    private System.Windows.Forms.Timer? _timer;
    private string _code = "";
    private string _apiBase = "";
    private int _polls;
    private bool _inflight;
    private Action<string, bool, bool>? _update;

    private static string WebOrigin(string apiBase) =>
        apiBase.EndsWith("/api", StringComparison.Ordinal) ? apiBase[..^4] : apiBase;

    /// update(message, done, success) — called on the UI thread.
    public void Connect(Action<string, bool, bool> update)
    {
        Cancel();
        _update = update;
        _apiBase = Settings.ApiBaseUrl;
        _code = Guid.NewGuid().ToString("N"); // 32 hex chars
        _polls = 0;

        var link = $"{WebOrigin(_apiBase)}/link?code={_code}";
        try { System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo(link) { UseShellExecute = true }); } catch { }
        _update?.Invoke("Opened your browser — sign in or create an account…", false, false);

        _timer = new System.Windows.Forms.Timer { Interval = 2500 };
        _timer.Tick += async (_, _) => await Poll();
        _timer.Start();
    }

    public void Cancel()
    {
        _timer?.Stop();
        _timer?.Dispose();
        _timer = null;
        _inflight = false;
    }

    private async Task Poll()
    {
        if (_inflight) return;
        if (++_polls > 120) { Cancel(); _update?.Invoke("Timed out — click Connect to try again.", true, false); return; }
        _inflight = true;
        try
        {
            using var resp = await Http.GetAsync($"{_apiBase}/auth/pair?code={_code}");
            if ((int)resp.StatusCode == 200)
            {
                var json = await resp.Content.ReadAsStringAsync();
                using var doc = JsonDocument.Parse(json);
                if (doc.RootElement.TryGetProperty("apiKey", out var k) && k.GetString() is { Length: > 0 } key)
                {
                    Settings.SaveApiKey(key);
                    Cancel();
                    _update?.Invoke("✓ Connected — live ads enabled.", true, true);
                    return;
                }
            }
            // 204 (pending) or transient error: keep waiting.
            _update?.Invoke("Waiting for you to finish in the browser…", false, false);
        }
        catch { /* transient — keep polling */ }
        finally { _inflight = false; }
    }
}

File: apps/overlay-macos/Sources/PrismAuth.m
#import "PrismAuth.h"
#import <Cocoa/Cocoa.h>

static const NSTimeInterval kPollInterval = 2.5;
static const int kMaxPolls = 120; // ~5 minutes

@interface PrismAuth ()
@property(nonatomic, copy) NSString *code;
@property(nonatomic, copy) NSString *apiBase;     // e.g. https://goprism.dev/api
@property(nonatomic, copy) NSString *webOrigin;   // e.g. https://goprism.dev
@property(nonatomic, strong) NSTimer *timer;
@property(nonatomic, assign) int polls;
@property(nonatomic, assign) BOOL inflight;
@property(nonatomic, copy) void (^update)(NSString *, BOOL, BOOL);
@end

@implementation PrismAuth

- (NSString *)resolveApiBase {
    NSString *u = [NSProcessInfo processInfo].environment[@"PRISM_API_URL"];
    if (!u.length) u = [[NSUserDefaults standardUserDefaults] stringForKey:@"PrismApiUrl"];
    if (!u.length) u = @"https://goprism.dev/api";
    if ([u hasSuffix:@"/"]) u = [u substringToIndex:u.length - 1];
    return u;
}

- (NSString *)originFromBase:(NSString *)base {
    if ([base hasSuffix:@"/api"]) return [base substringToIndex:base.length - 4];
    return base;
}

- (NSString *)randomCode {
    uint8_t bytes[16];
    arc4random_buf(bytes, sizeof(bytes));
    NSMutableString *s = [NSMutableString stringWithCapacity:32];
    for (size_t i = 0; i < sizeof(bytes); i++) [s appendFormat:@"%02x", bytes[i]];
    return s;
}

- (void)connectWithStatus:(void (^)(NSString *, BOOL, BOOL))update {
    [self cancel];
    self.update = update;
    self.apiBase = [self resolveApiBase];
    self.webOrigin = [self originFromBase:self.apiBase];
    self.code = [self randomCode];
    self.polls = 0;

    NSString *link = [NSString stringWithFormat:@"%@/link?code=%@", self.webOrigin, self.code];
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:link]];
    [self emit:@"Opened your browser — sign in or create an account…" done:NO success:NO];

    self.timer = [NSTimer scheduledTimerWithTimeInterval:kPollInterval target:self
                                                selector:@selector(poll) userInfo:nil repeats:YES];
}

- (void)cancel {
    [self.timer invalidate];
    self.timer = nil;
    self.inflight = NO;
}

- (void)emit:(NSString *)msg done:(BOOL)done success:(BOOL)success {
    if (!self.update) return;
    void (^cb)(NSString *, BOOL, BOOL) = self.update;
    dispatch_async(dispatch_get_main_queue(), ^{ cb(msg, done, success); });
}

- (void)poll {
    if (self.inflight) return;
    if (++self.polls > kMaxPolls) {
        [self cancel];
        [self emit:@"Timed out — click Connect to try again." done:YES success:NO];
        return;
    }
    self.inflight = YES;

    NSString *u = [NSString stringWithFormat:@"%@/auth/pair?code=%@", self.apiBase, self.code];
    NSMutableURLRequest *req = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:u]];
    req.timeoutInterval = 8.0;

    __weak PrismAuth *weakSelf = self;
    [[[NSURLSession sharedSession] dataTaskWithRequest:req
        completionHandler:^(NSData *data, NSURLResponse *resp, NSError *err) {
            PrismAuth *self = weakSelf;
            if (!self) return;
            self.inflight = NO;
            NSInteger status = [(NSHTTPURLResponse *)resp statusCode];
            if (status == 200 && data.length) {
                id json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
                NSString *key = [json isKindOfClass:[NSDictionary class]] ? json[@"apiKey"] : nil;
                if (key.length) {
                    [[NSUserDefaults standardUserDefaults] setObject:key forKey:@"PrismApiKey"];
                    [self cancel];
                    [self emit:@"✓ Connected — live ads enabled." done:YES success:YES];
                    return;
                }
            }
            // 204 (pending) or transient error: keep waiting.
            [self emit:@"Waiting for you to finish in the browser…" done:NO success:NO];
        }] resume];
}

@end
